src/node.rs

```diff
@@ -114,16 +126,263 @@ impl InMemoryNodeInner {
+
+            // In order for execution to pass smoothly, we need to ensure that block's required gasPerPubdata will be
+            // <= to the one in the transaction itself.
+            adjust_l1_gas_price_for_tx(
+                current_l1_gas_price,
+                L2_GAS_PRICE,
+                tx.gas_per_pubdata_byte_limit(),
+            )
+        };
+
+        let (base_fee, gas_per_pubdata_byte) =
+            derive_base_fee_and_gas_per_pubdata(l1_gas_price, fair_l2_gas_price);
+
+        // Check for properly formatted signature
+        if l2_tx.common_data.signature.is_empty() {
+            l2_tx.common_data.signature = vec![0u8; 65];
+            l2_tx.common_data.signature[64] = 27;
+        }
+
+        l2_tx.common_data.fee.gas_per_pubdata_limit = MAX_GAS_PER_PUBDATA_BYTE.into();
+        l2_tx.common_data.fee.max_fee_per_gas = base_fee.into();
+        l2_tx.common_data.fee.max_priority_fee_per_gas = base_fee.into();
+
+        let mut storage_view = StorageView::new(&self.fork_storage);
+
+        // Calculate gas_for_bytecodes_pubdata
+        let pubdata_for_factory_deps = l2_tx
+            .execute
+            .factory_deps
+            .as_deref()
+            .unwrap_or_default()
+            .iter()
+            .map(|bytecode| {
+                if storage_view.is_bytecode_known(&hash_bytecode(bytecode)) {
+                    return 0;
+                }
+
+                let length = if let Ok(compressed) = compress_bytecode(bytecode) {
+                    compressed.len()
+                } else {
+                    bytecode.len()
+                };
+                let publish_byte_overhead = 100;
+                length as u32 + publish_byte_overhead
+            })
+            .sum::<u32>();
+
+        if pubdata_for_factory_deps > MAX_PUBDATA_PER_BLOCK {
+            return Err(into_jsrpc_error(Web3Error::SubmitTransactionError(
+                "exceeds limit for published pubdata".into(),
+                Default::default(),
+            )));
+        }
+
+        let gas_for_bytecodes_pubdata: u32 =
+            pubdata_for_factory_deps * (gas_per_pubdata_byte as u32);
+
+        // We are using binary search to find the minimal values of gas_limit under which the transaction succeeds
+        let mut lower_bound = 0;
+        let mut upper_bound = MAX_L2_TX_GAS_LIMIT as u32;
+        let acceptable_overestimation = 1_000;
+        let max_attempts = 30usize;
+
+        let mut number_of_iterations = 0usize;
+        while lower_bound + acceptable_overestimation < upper_bound
+            && number_of_iterations < max_attempts
+        {
+            let mid = (lower_bound + upper_bound) / 2;
+            let try_gas_limit = gas_for_bytecodes_pubdata + mid;
+
+            let estimate_gas_result = self.estimate_gas_step(
+                l2_tx.clone(),
+                gas_per_pubdata_byte,
+                try_gas_limit,
+                l1_gas_price,
+                base_fee,
+            );
```

Comment:
If possible it would be good to refactor this so `estimate_gas_step` can take in a reference to `l2_tx` instead of cloning it every iteration.

Reply:
We need to clone as I'm modifying the transaction every iteration